### ISSUE TITLE
Support all RFC 4122 UUID versions

### ISSUE THREAD

**Issue by norguhtar** [tarantool/tarantool#5444]
Tarantool version: 2.3

OS version: linux

Bug description: uuid validate check not correct

Steps to reproduce:
```
tarantool> uuid = require("uuid")
tarantool> uuid.fromstr("bea80698-e07d-11ea-fe85-00155d373b0c")
---
- null
...

tarantool> uuid.fromstr("bee815b2-e07d-11ea-fe85-00155d373b0c")
---
- null
...
```

For example this uuid work in resty uuid 
```
uuid = require("resty.uuid")
uuid.is_valid("bea80698-e07d-11ea-fe85-00155d373b0c")
```
For example this uuid in python 3.7.9

```
>>> import uuid
>>> uuid.UUID("bee815b2-e07d-11ea-fe85-00155d373b0c");
UUID('bee815b2-e07d-11ea-fe85-00155d373b0c')
>>> uuid.UUID("bea80698-e07d-11ea-fe85-00155d373b0c");
UUID('bea80698-e07d-11ea-fe85-00155d373b0c')
```

**Comment by Totktonada** [comment]
```
bea80698-e07d-11ea-fe85-00155d373b0c
              ^^ time_hi_and_version = 0x11
```

Four most significant bits of `time_hi_and_version` are `0b0001`, so it is version 1 (generated from a timestamp), not version 4 (randomly generated). Tarantool supports only the version 4. I don't know why our [documentation](https://www.tarantool.io/en/doc/latest/reference/reference_lua/uuid/) have no any words about this, but the code states it cleanly:

https://github.com/tarantool/tarantool/blob/1ad696c09d5735afb0888e9351660b963f432f12/src/lib/uuid/tt_uuid.c#L56-L65

**Comment by norguhtar** [comment]
Really? Why this uuid worked?

```
tarantool> uuid = require("uuid")
---
...

tarantool> uuid.fromstr("4429d312-18d4-11eb-94f6-77e22d44915a")
---
- 4429d312-18d4-11eb-94f6-77e22d44915a
...

tarantool> uuid.fromstr("64ecacb4-18d4-11eb-94f6-cbb989f20a7e")
---
- 64ecacb4-18d4-11eb-94f6-cbb989f20a7e
...

tarantool> uuid.fromstr("64ecacb5-18d4-11eb-94f6-cf778123fd70")
---
- 64ecacb5-18d4-11eb-94f6-cf778123fd70
...

```

This first version i get from uuid_generate_v1mc() from PostgeSQL it worked fine. Why? :)

**Comment by Totktonada** [comment]
I guess because only variant value is verified, not version:

https://github.com/tarantool/tarantool/blob/1ad696c09d5735afb0888e9351660b963f432f12/src/lib/uuid/tt_uuid.h#L65-L73

There is no guarantee that a value generated by another algorithm will be declined. I guess we can make the validation more strict.

**Comment by norguhtar** [comment]
Why you try strict uuid to v4?

**Comment by Totktonada** [comment]
I don't know why it was implemented this way in the first place (I guess because it was enough), but now I just clarify how it works. Should we consider this issue as the feature request for support of other uuid versions? What is your use case?

**Comment by norguhtar** [comment]
Yes. Support all uuid versions it really strange support only 4 version.

**Comment by Totktonada** [comment]
'It is really strange' may be a reason of changes, but it usually works when a maintainer said this :) So I would kindly ask about your use case.

**Comment by norguhtar** [comment]
For example in postgresql when you use uuid as primary key and use v1 get more compact index. Also we can get other version from external system. For example 1C and Windows. When you use uuid internal all ok but when you try use uuid from external system you can get any version.

**Comment by Totktonada** [comment]
I renamed the issue to make it clear that it is the feature request. RFC 4122 describes versions 1, 3, 4, 5. The version 2 is mentioned, but is not described.

**Comment by imb590** [comment]
Is there a need to validate `uuid` values at all? What's wrong with treating any 16 byte value as `uuid`, like for example [PostgreSQL does](https://github.com/postgres/postgres/blob/REL_13_4/src/backend/utils/adt/uuid.c#L147)? The standard is subject to changes, and there are many things that generate `uuid`-like values that do not conform strictly to the existing standards. I'd rather like Tarantool just treated these values as opaque `uuid`s without trying to understand bits and pieces in them. The current behavior where you can send a value like `00000000-0000-0000-e000-000000000000` to Tarantool and it silently treats it as `nil` is very surprising.

**Comment by Totktonada** [comment]
There is a moment that should be considered when we'll implement generators for new UUID types.

The `uuid.new()` function reads from `/dev/urandom`, which returns PRNG values seeded from the entropy pool. (Or it uses the data from the entropy pool, when it is available, but falls back to PRNG otherwise? The manual does not descend into those details.). It looks as a slow way to calculate a next PRNG value (because it ia the syscall) and it possibly may be implemented entirely in a user space.

There are doubts regaring change it for the existing `uuid.new()` implementation: there may be unobvious drawbacks like less uniform or less entropy based distribution. But maybe we can try with the new UUID types.

Anyway, I noted it here just in the case: if we'll want to revisit this point, when we'll implement new UUID types.

**Comment by unera** [comment]
I think that we should drop any uuid validation excluding symbol checks.

Do not add validation, drop validation!

so
- invalid uuid contains non `[0-9a-fA-F]` symbols (and `-`)
- valid uuid can be built from string that skips `-` symbols, example:
   - `98c0cfc3-2b03-461d-b662-9a869bf46c75` - valid uuid
   - `98c0cfc32b03461db6629a869bf46c75` - valid uuid
   - `bea80698-e07d-11ea-fe85-00155d373b0c` - valid uuid


So, uuid is a datatype to store 128-bit value.

Let's think that UUID - is a 128bit integer. 
Someone has provided them as unique integers. 

We shouldn't think how has he done that. We should think how could the value be saved/restored.

**Comment by Mons** [comment]
Let's completely relax uuid v4 check (just remove validation) and accept arbitrary 128-bit sequence as an uuid.

**Comment by Totktonada** [comment]
I agree that it worth to allow arbitrary 16 byte *payload* for an UUID.

I would keep the validation/parsing of the *string representation* as defined in grammars in [RFC 4122, section 3](https://datatracker.ietf.org/doc/html/rfc4122#section-3) and [RFC 9562, section 4](https://datatracker.ietf.org/doc/html/rfc9562#name-uuid-format). If there are objections, for example, an alternative widespread string representation, I'm asking for details here.

To sum up: I propose to relax payload validation, but keep the string representation validation.

I also propose to consider the following points as out of scope of this issue:

* Constructors that generate UUIDs other than v4.
* Constructors that provide secure/fast generation.

**Comment by Totktonada** [comment]
> To be honest, it passed almost 4 years since this ticket and for now, it's worth adding support for uuid v7/v8

It has the same size and the same string representation format. Relaxing the payload validation effectively adds support for them. @Buristan has added testing values from RFC 9562 into tarantool testing in PR #10185.
